### ISSUE TITLE
Firmware 2.3 broke GB/GBC cartridge .sav files created via Memories

### ISSUE THREAD

**Issue by Blaziken257** [frederic-mahe/Analogue-Pocket#137]
### Description

In Analogue Pocket firmware 2.3, generating a save state when running a game on a cartridge results in a corrupted .sav file when attempting to copy the save to a computer (or another device) and running it on an emulator. Depending on the game, this can result in a "Save corrupted" error message or quietly have an empty save file.

**TLDR: The reason is that on this new firmware, the Pocket inserts a single random byte at the beginning of the save file, shifting all other bytes rightward by one. It also erases the last byte of the save file.**

This issue didn't exist in 2.2 or earlier. Even on 2.3, it doesn't seem to cause any issues if these saves are only used on the Pocket.

This issue was likely first reported in this Reddit thread: https://www.reddit.com/r/AnaloguePocket/comments/1fg7nt1/trouble_transferring_pok%C3%A9mon_gb_saves_to/

### Details and sample save files

An example of a corrupted Pokémon Red save from Analogue Pocket FW 2.3, and one from an Everdrive GB X7 (using the save state created from the cartridge, loading the state on the X7, and extracting the good save afterwards) is shown below in the following screenshots. Notice an extra `01` byte inserted on the right that doesn't belong:

| Everdrive GB X7 (FW 2.3, Valid save) | Analogue Pocket FW 2.3 (corrupted via Memories) |
| ------------- | ------------- |
| ![Everdrive GB Save Pokemon Red](https://github.com/user-attachments/assets/daf9664e-e7d3-4cdf-a151-1fc0cda9de3d) | ![Pocket Memories Save Pokemon Red](https://github.com/user-attachments/assets/a0f82a4c-de39-4dea-a592-3982b6da5e46)|

Attached at the end of this issue are several save files and save states generated through firmwares 2.2 and 2.3. The following games were used as examples: Pokémon Red (English), Donkey Kong Land III (English Rev A), and Super Mario Bros. Deluxe (English Rev B). These games were chosen to show that GB and GBC games have this issue, and to show that this happens with different SRAM sizes.

Both firmwares have a save and save state created via the Memories feature, and a save from the Everdrive GB X7. As stated before, the save state for each game was first created on a retail cartridge, then loaded on an Everdrive to create a save from that.

For 2.3 only, all of the saves in the "Analogue Memories" subfolder are corrupted, and attempting to run them on an emulator will result in an empty save (for Donkey Kong Land III and Super Mario Bros. Deluxe) or a save corruption message (for Pokémon Red).

The 2.3 folder also has a "Analogue - Fixed" folder where the corrupted files above are manually fixed. This was done by removing the offending first byte, and to insert an `FF` byte at the end. This happens to work for these games, but if there is a game that uses the last byte as part of a checksum, this method won't work.

In summary, this issue (where cart saves can't be properly extracted via Memories) can be worked around with the following methods:

- Downgrade to 2.2
- Create the save state on a retail cart, then load it on a flashcart
- Manually fix the corrupt save by removing the offending first byte in a .sav and inserting a byte at the end, and hope this works for every game
- Buy an external save backup device

None of these methods are ideal; I personally downgraded to 2.2 and hope this gets resolved in a future update.
[Analogue Pocket Saves.zip](https://github.com/user-attachments/files/17004528/Analogue.Pocket.Saves.zip)

**Comment by Blaziken257** [comment]
This appears to have finally been fixed in firmware 2.5.
